Changing user credentials | #390 Passwort / E-Mail-Adresse im Profil ändern

abatmas opened the pull request.

Commit:
Added a new column for the new_email_to_validate

Commit:
Updated User View, the user can now change the password

Commit:
Refactoring

SimonK92 commented on src/main/java/backend/view/user/UserView.kt:
```diff
@@ -12,7 +12,7 @@ import org.hibernate.validator.constraints.SafeHtml.WhiteListType.NONE
 import javax.validation.constraints.NotNull
 
-class UserView() {
+open class UserView() {
```
Was bedeutet das "open"?

sibbl replied:
Standardmäßig sind Klassen in Kotlin `final`, d.h. durch das `open` wird es möglich diese zu erschreiben und woanders um ein nur an dieser Stelle notwendiges Feld wie `newPassword` zu ergänzen.

SimonK92 commented:
@abatmas Vor Mergen bitte Sebastian um Review bitten.

sibbl commented on src/main/java/backend/controller/UserController.kt:
```diff
@@ -130,23 +127,48 @@ class UserController(private val userService: UserService,
         user.setValuesFrom(body)
-        userService.save(user)
 
+        userService.getUserByEmail(user.email)
```
Unnötigen Code entfernen

sibbl commented on src/main/java/backend/controller/UserController.kt:
```diff
@@ -130,23 +127,48 @@ class UserController(private val userService: UserService,
+     * Checks if the current password given from the user at the account setting is equal with the users password
+     */
+    private fun User.hasPassword(value: String?): Boolean {
+        if (BCryptPasswordEncoder().matches(value, this.passwordHash)){
```
Ohne if-else direkt den Rückgabewert returnen

sibbl commented on src/main/java/backend/controller/UserController.kt:
```diff
@@ -265,6 +287,7 @@ class UserController(private val userService: UserService,
         this.gender = userView.gender ?: this.gender
         this.profilePic = userView.profilePic?.let(::Media) ?: this.profilePic
         this.newsletter = userView.newsletter ?: this.newsletter
+        this.newEmailToValidate = userView.newEmailToValidate ?: this.newEmailToValidate
```
Hier fehlt komplett die Logik, dass dem Nutzer an die neue E-Mail-Adresse eine E-Mail zum Bestätigen der E-Mail-Adresse gesendet wird.

Dazu fehlt noch ein Frontend- und Backend-Endpunkt, wo diese Aktivierung dann erfolgt und die neue E-Mail-Adresse als die aktive gesetzt wird.

sibbl committed:
Finalize email+password change features

- add sending activation email
- change activation requests to POST
- move some methods to where they belong

sibbl committed:
Fix email change confirmation mail being sent to old email

sibbl commented:
@SimonAbat ich würde mich freuen wenn du nochmal drüber schauen kannst.

Ich hab jetzt noch den Versand der Passwort-Ändern-E-Mail nachgezogen, die Aktivierungs-Endpunkte ergänzt und das Ändern des Passworts noch ein wenig aufgehübscht.

sibbl committed:
Fix build

sibbl committed:
Fix build error
